### PR TITLE
Add a new host function for reporting fatal errors; make WASM backtraces readable when printing out errors

### PR DESCRIPTION
**Pull request by koute**
This PR adds a new host function for reporting fatal errors (for situations when the runtime either panics or runs out of memory) which are then passed through the executor, instead of being only printed out in the logs.

Issue: https://github.com/paritytech/substrate/issues/10585
Previous attempt: https://github.com/paritytech/substrate/pull/7954

A few notes regarding the implementation:
- Unlike the original PR here the host function has automatic trap semantics and will abort the execution when called. (The abort is triggered on the WASM side however, since the WASM FFI has no concept of functions which always trap AFAIK.)
- ~~I've tried to avoid the Rust-specific term "panic" when picking a name for it, instead opting for "fatal error". Naming is hard though, so if anyone has a better suggestion I'm all ears. (:~~ Changed back to "panic".
- I've added two new variants to the `Error` enum - one for normal WASM traps (`AbortedDueToTrap`), and one specifically for traps triggered through this host function (~~`AbortedDueToFatalError`~~`AbortedDueToPanic`). I've also removed the `wasmi`-specific `Error::Trap`, and now traps for both backends result in the same `Error` variant.
- The two new variants have a backtrace attached to them; from what I can see `wasmi` doesn't seem to support gathering backtraces though, so under `wasmi` it's always empty.
- I took advantage of `wasmtime` to serialize the backtrace into a `String` for us, [since doing that manually is not totally trivial](https://docs.rs/wasmtime/0.33.0/src/wasmtime/trap.rs.html#314-377). It can be considered a little hacky, but it works.
- I've used the new host function both in the panic handler as well as in the out of memory handler, however it is currently disabled by default, gated by a feature flag so that it's not used by default yet can still be tested. I'm guessing we probably want to first cut a new release of the host with the new host function exposed and only *then* enable it in our runtimes, right?
- The `primitives/application-crypto` crate depends on the the `sp-io` crate *without* the `std` feature, however it is **not** compiled for the `wasm32-unknown-unknown` target. This is problematic since the WASM-only interface which I've added calls `core::arch::wasm32::unreachable()`, which is available only on WASM targets, so this breaks `application-crypto`'s compilation. Not sure if there's a better way to handle it, but I've just made the `#[runtime_interface(wasm_only)]` *actually* be WASM-only by gating it with `#[cfg(target_arch = "wasm32")]`.

Future work/open questions:
- Do we want to add support for gathering backtraces to `wasmi` too?
- Do we want to decode the backtrace ourselves? (I'd say probably yes? Especially if we add backtraces from `wasmi`, to make sure they're consistent between both backends.)
- Do we want the backtrace to be accessible frame-by-frame instead of just being a `String` blob?
- Further cleanup of backend-specific error variants? (`wasmi` still has its own `Error::Wasmi` variant which we could get rid of)

### CODE REVIEW

**Comment by pepyakin** [comment]
- can you give an example how does an stack trace looks like? Do we need to enable the debug_info to make them useful?
- wasmi does not yet support backtraces and likely won't support them soon. Worth creating an issue though.

>"Do we want the backtrace to be accessible frame-by-frame instead of just being a String blob?"

Can't immediately think of a use-case. Perhaps can be followed up?

**Comment by koute** [comment]
> can you give an example how does an stack trace looks like? Do we need to enable the debug_info to make them useful?

In general since the backtraces are serialized into a string the same way as they were before (that is: `wasmtime` does the work) they look just like they did before, except now they are stored as their own field instead of being mashed together with the error message in the same `String`. Here's how a panic from the runtime looks like with this PR: (I've injected a `panic!` inside of the Kusama runtime, compiled whole `polkadot` in release mode, and ran that as a local testnet)

```
2022-01-31 23:23:30 Proposing failed: ClientImport("RuntimeApiError(Application(Execution(AbortedDueToFatalError(MessageWithBacktrace { message: \"panicked at 'dummy panic in runtime', /local/cpp/substrate/frame/babe/src/lib.rs:349:13\", backtrace: Some(Backtrace { backtrace_string: \"\\n    0: 0x34f1 - <unknown>!rust_begin_unwind\\n    1: 0x314e - <unknown>!core::panicking::panic_fmt::h8f11323637b4db3b\\n    2: 0x28993f - <unknown>!<(TupleElement0,TupleElement1) as frame_support::traits::hooks::OnFinalize<BlockNumber>>::on_finalize::h6941af0ed46021a0\\n    3: 0x2892e9 - <unknown>!frame_executive::Executive<System,Block,Context,UnsignedValidator,AllPalletsWithSystem,COnRuntimeUpgrade>::idle_and_finalize_hook::ha916b2075fa788bc\\n    4: 0x1d3554 - <unknown>!frame_executive::Executive<System,Block,Context,UnsignedValidator,AllPalletsWithSystem,COnRuntimeUpgrade>::finalize_block::h37b63c83bba0ed0f\\n    5: 0x1d3520 - <unknown>!<kusama_runtime::Runtime as sp_block_builder::runtime_decl_for_BlockBuilder::BlockBuilder<sp_runtime::generic::block::Block<sp_runtime::generic::header::Header<u32,sp_runtime::traits::BlakeTwo256>,sp_runtime::generic::unchecked_extrinsic::UncheckedExtrinsic<sp_runtime::multiaddress::MultiAddress<<<sp_runtime::MultiSignature as sp_runtime::traits::Verify>::Signer as sp_runtime::traits::IdentifyAccount>::AccountId,()>,kusama_runtime::Call,sp_runtime::MultiSignature,(frame_system::extensions::check_non_zero_sender::CheckNonZeroSender<kusama_runtime::Runtime>,frame_system::extensions::check_spec_version::CheckSpecVersion<kusama_runtime::Runtime>,frame_system::extensions::check_tx_version::CheckTxVersion<kusama_runtime::Runtime>,frame_system::extensions::check_genesis::CheckGenesis<kusama_runtime::Runtime>,frame_system::extensions::check_mortality::CheckMortality<kusama_runtime::Runtime>,frame_system::extensions::check_nonce::CheckNonce<kusama_runtime::Runtime>,frame_system::extensions::check_weight::CheckWeight<kusama_runtime::Runtime>,pallet_transaction_payment::ChargeTransactionPayment<kusama_runtime::Runtime>)>>>>::finalize_block::h38f2d4aad71c02fd\\n    6: 0x2a7079 - <unknown>!BlockBuilder_finalize_block\\n\" }) }))))")
```

So,

1) The error message in this particular case is printed through `Debug`, so it's not very readable. We should probably switch it to use `Display`. (If I'd temporarily remove the `Debug` impl I'm guessing every such place should simply fail to compile and then we could just do a `{:?}` -> `{}` substitution; I'll take a look at it tomorrow.)
2) The way backtrace itself is serialized into a string (assuming it'd get printed with proper newlines) is okay-ish; I could make it more pretty if we'd do the decoding ourselves.
3) The function names are there, so the stack trace is useful, however there are no filenames nor line numbers. That requires extra debugging info which - I assume - we don't want to actually be a part of the runtime? But I guess we could maybe do something like this: build a runtime with full debugging symbols, strip them out for the blob we keep on chain, and then for the most recent runtime embed the blob with full debugging symbols inside of the `polkadot` binary and have it swap out the on-chain blob with the blob with extra symbols if it detects that it matches. Not sure if we actually want to do this though?

> wasmi does not yet support backtraces and likely won't support them soon. Worth creating an issue though.

Indeed that's how it looked like to me in the code. (:

I was just curious whether it is something that would be actually useful to us and we'd want it sooner rather than later, or would it be just a nice thing to have but we don't really care that much right now? (If it's the former I suppose I could add the support myself to `wasmi` as a followup.)

> > "Do we want the backtrace to be accessible frame-by-frame instead of just being a String blob?"
> 
> Can't immediately think of a use-case. Perhaps can be followed up?

Yeah, that was more of a "if anyone thinks it's needed I can add it, but otherwise I'll leave it as-is for now".

**Comment by koute** [comment]
Okay, so this was a huge pain, but I went through the whole codebase and everywhere an `Error` from an executor is printed (either directly or indirectly) I made sure that it's printed through `Display`, so now the backtraces should actually be readable(ish).

Before:

```
2022-01-31 23:23:30 Proposing failed: ClientImport("RuntimeApiError(Application(Execution(AbortedDueToFatalError(MessageWithBacktrace { message: \"panicked at 'dummy panic in runtime', /local/cpp/substrate/frame/babe/src/lib.rs:349:13\", backtrace: Some(Backtrace { backtrace_string: \"\\n    0: 0x34f1 - <unknown>!rust_begin_unwind\\n    1: 0x314e - <unknown>!core::panicking::panic_fmt::h8f11323637b4db3b\\n    2: 0x28993f - <unknown>!<(TupleElement0,TupleElement1) as frame_support::traits::hooks::OnFinalize<BlockNumber>>::on_finalize::h6941af0ed46021a0\\n    3: 0x2892e9 - <unknown>!frame_executive::Executive<System,Block,Context,UnsignedValidator,AllPalletsWithSystem,COnRuntimeUpgrade>::idle_and_finalize_hook::ha916b2075fa788bc\\n    4: 0x1d3554 - <unknown>!frame_executive::Executive<System,Block,Context,UnsignedValidator,AllPalletsWithSystem,COnRuntimeUpgrade>::finalize_block::h37b63c83bba0ed0f\\n    5: 0x1d3520 - <unknown>!<kusama_runtime::Runtime as sp_block_builder::runtime_decl_for_BlockBuilder::BlockBuilder<sp_runtime::generic::block::Block<sp_runtime::generic::header::Header<u32,sp_runtime::traits::BlakeTwo256>,sp_runtime::generic::unchecked_extrinsic::UncheckedExtrinsic<sp_runtime::multiaddress::MultiAddress<<<sp_runtime::MultiSignature as sp_runtime::traits::Verify>::Signer as sp_runtime::traits::IdentifyAccount>::AccountId,()>,kusama_runtime::Call,sp_runtime::MultiSignature,(frame_system::extensions::check_non_zero_sender::CheckNonZeroSender<kusama_runtime::Runtime>,frame_system::extensions::check_spec_version::CheckSpecVersion<kusama_runtime::Runtime>,frame_system::extensions::check_tx_version::CheckTxVersion<kusama_runtime::Runtime>,frame_system::extensions::check_genesis::CheckGenesis<kusama_runtime::Runtime>,frame_system::extensions::check_mortality::CheckMortality<kusama_runtime::Runtime>,frame_system::extensions::check_nonce::CheckNonce<kusama_runtime::Runtime>,frame_system::extensions::check_weight::CheckWeight<kusama_runtime::Runtime>,pallet_transaction_payment::ChargeTransactionPayment<kusama_runtime::Runtime>)>>>>::finalize_block::h38f2d4aad71c02fd\\n    6: 0x2a7079 - <unknown>!BlockBuilder_finalize_block\\n\" }) }))))")
```

After:

```
2022-02-03 16:46:30 Proposing failed: Import failed: Error at calling runtime api: Execution failed: Execution aborted due to fatal error: panicked at 'dummy panic', /local/cpp/substrate/frame/babe/src/lib.rs:349:13
WASM backtrace:

    0: 0x34f1 - <unknown>!rust_begin_unwind
    1: 0x314e - <unknown>!core::panicking::panic_fmt::h8f11323637b4db3b
    2: 0x28993f - <unknown>!<(TupleElement0,TupleElement1) as frame_support::traits::hooks::OnFinalize<BlockNumber>>::on_finalize::h6941af0ed46021a0
    3: 0x2892e9 - <unknown>!frame_executive::Executive<System,Block,Context,UnsignedValidator,AllPalletsWithSystem,COnRuntimeUpgrade>::idle_and_finalize_hook::ha916b2075fa788bc
    4: 0x1d3554 - <unknown>!frame_executive::Executive<System,Block,Context,UnsignedValidator,AllPalletsWithSystem,COnRuntimeUpgrade>::finalize_block::h37b63c83bba0ed0f
    5: 0x1d3520 - <unknown>!<kusama_runtime::Runtime as sp_block_builder::runtime_decl_for_BlockBuilder::BlockBuilder<sp_runtime::generic::block::Block<sp_runtime::generic::header::Header<u32,sp_runtime::traits::BlakeTwo256>,sp_runtime::generic::unchecked_extrinsic::UncheckedExtrinsic<sp_runtime::multiaddress::MultiAddress<<<sp_runtime::MultiSignature as sp_runtime::traits::Verify>::Signer as sp_runtime::traits::IdentifyAccount>::AccountId,()>,kusama_runtime::Call,sp_runtime::MultiSignature,(frame_system::extensions::check_non_zero_sender::CheckNonZeroSender<kusama_runtime::Runtime>,frame_system::extensions::check_spec_version::CheckSpecVersion<kusama_runtime::Runtime>,frame_system::extensions::check_tx_version::CheckTxVersion<kusama_runtime::Runtime>,frame_system::extensions::check_genesis::CheckGenesis<kusama_runtime::Runtime>,frame_system::extensions::check_mortality::CheckMortality<kusama_runtime::Runtime>,frame_system::extensions::check_nonce::CheckNonce<kusama_runtime::Runtime>,frame_system::extensions::check_weight::CheckWeight<kusama_runtime::Runtime>,pallet_transaction_payment::ChargeTransactionPayment<kusama_runtime::Runtime>)>>>>::finalize_block::h38f2d4aad71c02fd
    6: 0x2a7079 - <unknown>!BlockBuilder_finalize_block
```

From my point of view this should be good to go, pending any further review issues or CI issues.

If we want to also have filenames + line numbers in the backtrace that could be done with e.g. something like I described in my previous comment, although *if* we'd go for that I'd suggest to follow up in another PR.

**Comment by wigy-opensource-developer** [comment]
Forgive me my ignorance, but can you confirm that these changes do not allow new instances of non-atomic state changes? So whenever processing an extrinsic had a side-effect on the state, it still either fully done or no effect is done?

**Comment by koute** [comment]
> Forgive me my ignorance, but can you confirm that these changes do not allow new instances of non-atomic state changes? So whenever processing an extrinsic had a side-effect on the state, it is still either fully done or no effect is done?

AFAIK that's how it works, and this PR doesn't change anything in this regard.

File: client/executor/common/src/error.rs
Comment (kianenigma) [inline client/executor/common/src/error.rs:174]: Does this mean it might exist? I'd call it 'maybe_backtrace' then.

Comment (koute) [reply client/executor/common/src/error.rs:174]: Hm... I *think* idiomatically people don't usually prepend `maybe_` on `Option` fields in Rust since the type is there anyway to signify that it's optional? At least I don't see it done often. And ideally we'd like this to be not an `Option`; it's only an `Option` because `wasmi` doesn't currently support backtraces.

But I don't really care *either* way, so I'm fine with renaming it if you want.

File: client/executor/wasmtime/src/instance_wrapper.rs
Comment (bkchr) [inline client/executor/wasmtime/src/instance_wrapper.rs:83]: Some comment on what we are removing here would have been nice.

File: primitives/runtime-interface/proc-macro/src/runtime_interface/bare_function_interface.rs
Comment (bkchr) [inline primitives/runtime-interface/proc-macro/src/runtime_interface/bare_function_interface.rs:107]: Why not require that the declaration directly uses `!` as return type? Would make it less "magic".

If not, we could also use here `parse_quote!( -> ! )`?

File: primitives/runtime-interface/proc-macro/src/utils.rs
Comment (bkchr) [inline primitives/runtime-interface/proc-macro/src/utils.rs:73]: Yeah, I would directly make the declaration return the never type.

File: primitives/runtime-interface/proc-macro/src/runtime_interface/bare_function_interface.rs
Comment (koute) [reply primitives/runtime-interface/proc-macro/src/runtime_interface/bare_function_interface.rs:107]: I actually considered that too, but then it kinda becomes magic in a different way... (:

For example:

```rust
fn foobar(&mut self) -> ! {
    do_stuff();
}
```

...if this was a normal Rust function it wouldn't compile because it returns `()` and *not* `!`. So in a way it's confusing because the function is then *not* supposed to actually return what its prototype says it returns! (Since the body is executed *on the host's side* while the `core::arch::wasm32::unreachable` is executed *on the WASM's side*.)

Alternatively to resolve this inconsistency we could:

a) Require that the host function actually *does* return `!`; that'd probably require a `panic!` inside, and for us to catch it at the boundary, but that seems kinda.... silly? (And also, `!` doesn't necessarily mean panic; it can also mean e.g. `std::process::exit` so conceptually it doesn't perfectly translate to the concept of a trap.)
b) Require that the host function returns a zero-sized sentinel type instead of `!`, but that just seems like more work for not much gain.

So in the end I went the attribute route. Not sure if there's another better alternative that I'm missing?

**Review by pepyakin — APPROVED** [review]
This looks good. As always nits are not necessary to fix in this PR and can be followed-up.

A word on the naming though. I must say I am not really happy with: fatal error either. In hindsight maybe "panic" is not that bad: maybe it is tolerable that it is taken from Rust, maybe it's time to admit we are rust-first ) Moreover, IIUC, Go uses the same term.

This all are just thoughts: by no means the naming blocks this PR.

File: primitives/runtime-interface/proc-macro/src/runtime_interface/bare_function_interface.rs
Comment (bkchr) [reply primitives/runtime-interface/proc-macro/src/runtime_interface/bare_function_interface.rs:107]: > ...if this was a normal Rust function it wouldn't compile because it returns `()` and _not_ `!`

Good argument. Whatever we do, it will be a little bit weird :D Fine to keep it the current way.

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
> In hindsight maybe "panic" is not that bad: maybe it is tolerable that it is taken from Rust, maybe it's time to admit we are rust-first ) Moreover, IIUC, Go uses the same term.

Hahaha!

Then I would go with panic :P

**Comment by koute** [comment]
Okay, I changed it back to "panic". (:

I've addressed all of your comments; should be good to go once the tests pass unless there's anything else you'd like me to change.

**Comment by bkchr** [comment]
> Okay, I changed it back to "panic". (:
> 
> I've addressed all of your comments; should be good to go once the tests pass unless there's anything else you'd like me to change.

I think the only think would be to create a ticket in Polkadot to enable this feature in like 2-3 versions. IMO we can then also remove the feature entirely.

**Comment by koute** [comment]
bot merge

File: primitives/io/src/lib.rs
Comment (tomaka) [inline primitives/io/src/lib.rs:1299]: Why not call it `abort`? The host interface isn't concerned with the fact that a panic happened, in fact it shouldn't even concerned with the fact that the runtime is written in Rust.

Comment (koute) [reply primitives/io/src/lib.rs:1299]: Hm, well, we can still change it. I don't really have a strong opinion *either* way and don't really want to get into bikeshedding the name, so I'm fine with whatever's the consensus.

**Pull request merged**
